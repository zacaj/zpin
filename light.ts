import { dImage, DisplayContent } from "./disp";

export enum Color {
    Red = 'Red',
    Green = 'Green',
    White = 'White',
    Orange = '#FF3000',
    Yellow = 'Yellow',
    Blue = '#0000A0',
    Purple = '#230054',
    Pink = '#ff0066',
    Black = 'Black',
    Gray = '#393939',
}

export type Frequency = number;

export type NormalizedLight = {
    color: Color;
    type: 'solid'|'pulsing'|'flashing';
    freq: Frequency;
    phase: number;
};

export type LightState = Color|{
    color: Color;
    flashing: true;
    freq: Frequency;
    phase: number;
}|{
    color: Color;
    pulsing: true;
    freq: Frequency;
    phase: number;
}|
NormalizedLight
|[Color|undefined, 'fl'|'flashing'|'pl'|'pulsing'|'flash'|'pulse'|false|undefined, Frequency?, number?]
|undefined|false;
const defaultFlashFreq: Frequency = 3;
const defaultPulseFreq: Frequency = 1;
export function normalizeLight(state: LightState): NormalizedLight|undefined {
    if (!state) return undefined;
    if (typeof state === 'string')
        return {
            color: state,
            type: 'solid',
            freq: defaultFlashFreq,
            phase: 0,
        };
    if (Array.isArray(state)) {
        if (!state[0]) return undefined;
        return {
            color: state[0],
            type: typeof state[1]==='string'? (state[1].startsWith('f')? 'flashing' : 'pulsing') : 'solid',
            freq: state[2] ?? (typeof state[1]==='string'&&state[1]?.startsWith('f')? defaultFlashFreq : defaultPulseFreq),
            phase: state[3] ?? 0,
        };
    }
    if ('type' in state) return state;
    return {
        color: state.color,
        type: 'flashing' in state? 'flashing' : 'pulsing',
        freq: state.freq ?? ('flashing' in state? defaultFlashFreq : defaultPulseFreq),
        phase: state.phase ?? 0,
    };
}

export function colorToHex(color: Color): string|undefined {
    if (typeof color === "string" && color.startsWith("#"))
        return color;
    const colors = {'aliceblue':'#f0f8ff','antiquewhite':'#faebd7','aqua':'#00ffff','aquamarine':'#7fffd4','azure':'#f0ffff',
    'beige':'#f5f5dc','bisque':'#ffe4c4','black':'#000000','blanchedalmond':'#ffebcd','blue':'#0000ff','blueviolet':'#8a2be2','brown':'#a52a2a','burlywood':'#deb887',
    'cadetblue':'#5f9ea0','chartreuse':'#7fff00','chocolate':'#d2691e','coral':'#ff7f50','cornflowerblue':'#6495ed','cornsilk':'#fff8dc','crimson':'#dc143c','cyan':'#00ffff',
    'darkblue':'#00008b','darkcyan':'#008b8b','darkgoldenrod':'#b8860b','darkgray':'#a9a9a9','darkgreen':'#006400','darkkhaki':'#bdb76b','darkmagenta':'#8b008b','darkolivegreen':'#556b2f',
    'darkorange':'#ff8c00','darkorchid':'#9932cc','darkred':'#8b0000','darksalmon':'#e9967a','darkseagreen':'#8fbc8f','darkslateblue':'#483d8b','darkslategray':'#2f4f4f','darkturquoise':'#00ced1',
    'darkviolet':'#9400d3','deeppink':'#ff1493','deepskyblue':'#00bfff','dimgray':'#696969','dodgerblue':'#1e90ff',
    'firebrick':'#b22222','floralwhite':'#fffaf0','forestgreen':'#228b22','fuchsia':'#ff00ff',
    'gainsboro':'#dcdcdc','ghostwhite':'#f8f8ff','gold':'#ffd700','goldenrod':'#daa520','gray':'#808080','green':'#00F000','greenyellow':'#adff2f',
    'honeydew':'#f0fff0','hotpink':'#ff69b4',
    'indianred ':'#cd5c5c','indigo':'#4b0082','ivory':'#fffff0','khaki':'#f0e68c',
    'lavender':'#e6e6fa','lavenderblush':'#fff0f5','lawngreen':'#7cfc00','lemonchiffon':'#fffacd','lightblue':'#add8e6','lightcoral':'#f08080','lightcyan':'#e0ffff','lightgoldenrodyellow':'#fafad2',
    'lightgrey':'#d3d3d3','lightgreen':'#90ee90','lightpink':'#ffb6c1','lightsalmon':'#ffa07a','lightseagreen':'#20b2aa','lightskyblue':'#87cefa','lightslategray':'#778899','lightsteelblue':'#b0c4de',
    'lightyellow':'#ffffe0','lime':'#00ff00','limegreen':'#32cd32','linen':'#faf0e6',
    'magenta':'#ff00ff','maroon':'#800000','mediumaquamarine':'#66cdaa','mediumblue':'#0000cd','mediumorchid':'#ba55d3','mediumpurple':'#9370d8','mediumseagreen':'#3cb371','mediumslateblue':'#7b68ee',
    'mediumspringgreen':'#00fa9a','mediumturquoise':'#48d1cc','mediumvioletred':'#c71585','midnightblue':'#191970','mintcream':'#f5fffa','mistyrose':'#ffe4e1','moccasin':'#ffe4b5',
    'navajowhite':'#ffdead','navy':'#000080',
    'oldlace':'#fdf5e6','olive':'#808000','olivedrab':'#6b8e23','orange':'#ffa500','orangered':'#ff4500','orchid':'#da70d6',
    'palegoldenrod':'#eee8aa','palegreen':'#98fb98','paleturquoise':'#afeeee','palevioletred':'#d87093','papayawhip':'#ffefd5','peachpuff':'#ffdab9','peru':'#cd853f','pink':'#FF00CC','plum':'#dda0dd','powderblue':'#b0e0e6','purple':'#800080',
    'rebeccapurple':'#663399','red':'#ff0000','rosybrown':'#bc8f8f','royalblue':'#4169e1',
    'saddlebrown':'#8b4513','salmon':'#fa8072','sandybrown':'#f4a460','seagreen':'#2e8b57','seashell':'#fff5ee','sienna':'#a0522d','silver':'#c0c0c0','skyblue':'#87ceeb','slateblue':'#6a5acd','slategray':'#708090','snow':'#fffafa','springgreen':'#00ff7f','steelblue':'#4682b4',
    'tan':'#d2b48c','teal':'#008080','thistle':'#d8bfd8','tomato':'#ff6347','turquoise':'#40e0d0',
    'violet':'#ee82ee',
    'wheat':'#f5deb3','white':'#ffffff','whitesmoke':'#f5f5f5',
    'yellow':'#ffff00','yellowgreen':'#9acd32'};

    if ((colors as any)[color.toLowerCase()])
        return (colors as any)[color.toLowerCase()];
    Log.error('console', 'no hex found for color "%s"', color);

    return undefined;
}

export function light(cond: boolean, color: LightState = Color.White, offColor?: LightState): LightState[] {
    return cond? [color] : (offColor? [offColor] : []);
}
export function flash(cond: boolean, color = Color.White, freqOrColor: Frequency|Color = defaultFlashFreq, freq = defaultFlashFreq, phase = 0): LightState[] {
    if (typeof freqOrColor === 'number')
        freq = freqOrColor;
    return cond? [{
        color,
        flashing: true,
        freq,
        phase,
    }] : (typeof freqOrColor !== 'number'? [[freqOrColor, 'flash', freq]] : []);
}
export function flashLight(cond: boolean, color = Color.White, off = Color.Red, freq = defaultFlashFreq, phase = 0): LightState[] {
    return cond? [{
        color,
        flashing: true,
        freq,
        phase,
    }] : [off];
}

export function colorToArrow(color?: Color): DisplayContent|undefined {
    if (!color) return undefined;
    return dImage(Object.entries(Color).find(e => e[1] === color)![0].toLowerCase()+'Arrow');
}

export function mix(func: (() => LightState|undefined)|LightState|undefined): (state?: LightState[]) => LightState[] | undefined {
    return prev => {
        const l = typeof func==='function'? func() : func;
        if (l)
            return [l, ...(prev??[])];
        else
            return prev;
    };
}

export function add(func: (() => boolean)|boolean, color: LightState|(() => LightState)): (state?: LightState[]) => LightState[] | undefined {
    return mix(() => (typeof func==='function'? func() : func)? (typeof color==='function'? color() : color) : undefined);
}

export function many(func: () => [LightState|'prev', boolean][]): (state?: LightState[]) => LightState[] | undefined {
    return (prev) => {
        const o = func();
        let state: LightState[] = [];
        for (const [key, value] of o) {
            if (key === 'prev' && value)
                state = [...state, ...prev ?? []];
            else if (value) {
                state.push(key as Color);
            }
        }
        return state;
    };
}

import {Socket} from 'net';
import { Log } from './log';
import { Image } from "./gfx";
import { Light, machine } from "./machine";
import { fork } from "./promises";
import { wait } from "./timer";
import { split, num } from "./util";

const socket = new Socket();

export const LPU = {
    isConnected: false,
    
    ip: '127.0.0.1',

    timeout: 10,
    connecting: false,

    promises: [] as ((resp: string) => void)[],

    async init(ip?: string) {
        // this.ip = '192.168.2.11';
        if (ip) this.ip = ip;

        socket.setTimeout(1000);
        socket.setNoDelay(true);
        socket.on('error', err => {
            Log.error(['lpu', 'console'],'socket error: ', err);
        });
        socket.on('close', err => {
            this.isConnected = false;
            this.connecting = false;
            Log.error(['lpu', 'console'], 'lost connection to LPU', err);
            this.reconnect();
        });
        this.connect();
    },

    reconnect() {
        if (this.connecting) return;
        this.connecting = true;
        fork(wait(this.timeout*1000).then(() => this.connect()));
        Log.info('lpu', 'reconnect to LPU in %is', this.timeout);
        if (this.timeout < 30)
            this.timeout *= 2;
    },

    connect() {
        if (this.isConnected) return;
        this.connecting = true;
        Log.info(['lpu'], 'connecting to %s', this.ip);
        socket.connect(9999, this.ip, async () => {
            Log.info(['lpu'],'reached LPU');
            this.isConnected = true;
            this.connecting = false;
            this.timeout = 10;

            await this.sendCommand("reset");
            await this.sendCommand(`setup 1,${Math.max(0, ...machine.lights.flatMap(l => l.nums ?? [])) + 1},3`);
            await this.sendCommand('init');
            await this.sendCommand('fill 1,0');
            await this.sendCommand(`thread_start 63, 0
    do
        delay 30
        render
    loop
thread_stop\n\n`);

            Log.log(['lpu', 'console'], 'connected to LPU');

            for (const light of machine.lights) {
                await light.sync();
            }
        });
    },

    async sendCommand(cmd: string, expectedResp: string|null = "", force = false): Promise<void> {
        if (!this.isConnected && !force) {
            Log.info('lpu', 'ignoring command %s, not connected', cmd);
            return;// "fake";
        }
        Log.info('lpu', 'send command %s', cmd);

        socket.write(cmd+(cmd.endsWith('\n')? '' : '\n'));
    },
}