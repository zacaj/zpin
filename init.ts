/* eslint-disable @typescript-eslint/no-floating-promises */
import { MPU } from './mpu';
import { machine, resetMachine } from './machine';
import { resetSwitchMatrix } from './switch-matrix';
import { Events } from './events';
import { Log } from './log';
import { initGfx } from './gfx';
import { Game } from './game';
import { initRecording, playRecording } from './recording';
import { wait } from './timer';
import { initAudio } from './sound';

const argv = require('yargs').argv;

export async function initMachine(mpu = true, gfx = false, game = false, trace = true, recording?: string, toPoint?: string) {
    try {
        if (argv.mpu !== undefined) mpu = argv.mpu;
        if (argv.gfx !== undefined) gfx = argv.gfx;
        if (argv.game !== undefined) game = argv.game;
        if (argv.trace !== undefined) trace = argv.trace;
        if (argv.recording !== undefined) recording = argv.recording;
        const sound = argv.sound ?? mpu;
        Log.init(trace);
        Log.log(['console'], 'Initializing....');
        Events.resetAll();
        resetSwitchMatrix();
        resetMachine();
        if (recording) {
            initRecording(recording);
        }
        if (mpu) {
            await MPU.init(argv.ip);
        
            await machine.initOutputs();
        }

        if (sound)
            await initAudio();
        if (gfx)
            await initGfx();
        if (game)
            Game.start();
        if (recording) {
            await new Promise(r => setTimeout(r, 100));
            await playRecording(toPoint);
        }
    } catch (err) {
        console.error('init error', err);
        debugger;
    }
}


if (require.main === module) {
    initMachine(!!argv.mpu, !!argv.gfx, !!argv.game);
}