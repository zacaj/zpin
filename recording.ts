import { Log } from './log';
import { Timer, passTime, setTime, time, wait } from './timer';
import * as fs from 'fs';
import { split, assert, debugging } from './util';
import { matrix } from './switch-matrix';
import { settleForks } from './promises';
import { MPU } from './mpu';
import { initMachine } from './init';
import { Events } from './events';
import { machine } from './machine';

export let curRecording: string|undefined = undefined;
let lines: string[];
let curTime = 0;
let curLine = 0;

export function initRecording(recording: string) {
    curRecording = recording;
    Log.log(['console'], 'playing back recording %s', recording);
    Timer.mockTime = 0;
    const file = fs.readFileSync(curRecording!, 'utf8');
    lines = file.split('\n');
    curTime = 0;
    curLine = 1;
}

export async function playRecording(toPoint?: string) {
    // await new Promise(r => setTimeout(r, 10000));
    const events = Events;
    const timer = Timer;
    const _machine = machine;
    for (; curLine < lines.length; curLine++) {
        const line = lines[curLine];
        if (line.startsWith('#')) continue;
        if (line && !line.includes(' ')) {
            const bp = line;
            if (bp === 'debug' && debugging()) {
                const nextLine = lines[curLine+1];
                debugger;
            } else if (process.env.NODE_ENV === 'test') {
                assert(bp === toPoint, `expected breakpoint ${toPoint} but got '${bp}'`);
                curLine++;
                await settleForks();
                return;
            }
            continue;
        }
        const [ts, sw, stateS, colS, rowS, timeS] = split(line, ' switch \'', '\' state -> ', ' (', ',', ') at ');
        if (timeS === undefined) continue;

        const time = parseInt(timeS, 10);
        const diff = time - curTime;
        assert(diff>=0, 'time travel not implemented');
        if (diff)
            await passTime(diff);
        curTime = time;

        matrix[parseInt(colS, 10)][parseInt(rowS, 10)]!.changeState(stateS === 'true', 'recording '+curRecording);
        await settleForks();
    }
    await settleForks();

    if (process.env.NODE_ENV !== 'test')
        await setTime(undefined);
    Log.log(['console', 'switch'], 'recording %s finished at %i', curRecording, time());
    curRecording = undefined;
}


export async function testRecording(name: string, toPoint?: string) {
    jest.spyOn(Log, 'init').mockImplementation(() => {});
    jest.spyOn(MPU, 'sendCommandCode').mockImplementation(async (cmd) => {
        return {
            code: 200,
            resp: 'mocked',
        };
    });
    await initMachine(false, false, true, false, `./recordings/${name}.rec`, toPoint);
}

export async function continueRecording(toPoint?: string) {
    await playRecording(toPoint);
}
export async function finishRecording() {
    await playRecording();
}
