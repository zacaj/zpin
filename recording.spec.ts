import { initRecording, playRecording, testRecording } from './recording';
import { matrix } from './switch-matrix';
import { machine } from './machine';
import { Color } from './light';
import { initMachine } from './init';
import { Log } from './log';
import { MPU } from './mpu';

describe('recordings', () => {
    test('reading', async () => {
        const shooter = jest.spyOn(matrix[0][0]!, 'changeState');
        const right4 = jest.spyOn(matrix[2][2]!, 'changeState');
        const right3 = jest.spyOn(matrix[3][2]!, 'changeState');
        const upper3a = jest.spyOn(matrix[2][5]!, 'changeState');
        expect(matrix[3][2]?.name).toBe('right 3');

        initRecording('./recordings/test-1.rec');
        await playRecording();

        expect(shooter).lastCalledWith(true, expect.stringContaining('recording'));
        expect(shooter).lastReturnedWith(expect.objectContaining({when: 2503}));
        expect(right4).lastCalledWith(true, expect.stringContaining('recording'));
        expect(right4).lastReturnedWith(expect.objectContaining({when: 7319}));
        expect(right3).lastCalledWith(true, expect.stringContaining('recording'));
        expect(right3).lastReturnedWith(expect.objectContaining({when: 8836}));
        expect(upper3a).lastCalledWith(true, expect.stringContaining('recording'));
        expect(upper3a).lastReturnedWith(expect.objectContaining({when: 13093}));
    });
    test('play test', async () => {
        await testRecording('lightStraight');
        expect(machine.out!.treeValues.lEjectStartMode).toEqual([Color.White]);
    });
});