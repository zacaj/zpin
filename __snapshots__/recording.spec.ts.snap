// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`recordings play test 1`] = `
Object {
  "outputs": Object {
    "centerBank": false,
    "getSkillshot": [Function],
    "iCenter1": "",
    "iCenter2": "",
    "iCenter3": "",
    "iLeft1": "",
    "iLeft2": "",
    "iLeft3": "",
    "iLeft4": "",
    "iMini1": "",
    "iMini2": "",
    "iMini3": "",
    "iRight1": "",
    "iRight2": "",
    "iRight3": "",
    "iRight4": "",
    "iRight5": "",
    "iSS1": "",
    "iSS2": "",
    "iSS3": "",
    "iSS4": "",
    "iSS5": "",
    "iSS6": "",
    "iSpinner": "",
    "iUpper21": "",
    "iUpper22": "",
    "iUpper31": "",
    "iUpper32": "",
    "iUpper33": "",
    "ignoreSkillsot": Set {},
    "kickerEnable": true,
    "lEjectArrow": Array [],
    "lEjectShowCards": Array [
      "White",
    ],
    "lEjectStartMode": Array [],
    "lLaneUpper1": Array [
      "Orange",
    ],
    "lLaneUpper2": Array [
      "Orange",
    ],
    "lLaneUpper3": Array [
      "Orange",
    ],
    "lLaneUpper4": Array [
      "Orange",
    ],
    "lLeftArrow": Array [],
    "lLowerRamp": Array [],
    "lMagnet1": Array [
      "Green",
    ],
    "lMagnet2": Array [
      "Green",
    ],
    "lMagnet3": Array [],
    "lMagnet4": Array [],
    "lMainTargetArrow": Array [],
    "lMiniReady": Array [
      "Red",
    ],
    "lPopperStatus": Array [
      "Green",
    ],
    "lPower1": Array [
      "Orange",
    ],
    "lPower2": Array [
      "Orange",
    ],
    "lPower3": Array [],
    "lPower4": Array [],
    "lRampArrow": Array [],
    "lRampMini": Array [
      "Orange",
    ],
    "lRampShowCards": Array [
      "White",
    ],
    "lRampStartMb": Array [
      Array [
        "Green",
        "fl",
      ],
    ],
    "lShooterLaneArrow": Array [],
    "lShooterShowCards": Array [
      "White",
    ],
    "lShooterStartHand": Array [
      Array [
        "Green",
        "fl",
      ],
    ],
    "lSideShotArrow": Array [],
    "lSideTargetArrow": Array [],
    "lSpinnerArrow": Array [],
    "lUpperLaneArrow": Array [],
    "lUpperTargetArrow": Array [],
    "leftBank": false,
    "leftGate": false,
    "leftMagnet": false,
    "lockPost": false,
    "magnetPost": false,
    "miniBank": false,
    "miniDiverter": false,
    "miniEject": false,
    "miniFlipperEnable": false,
    "outhole": false,
    "popper": false,
    "rampUp": false,
    "realRightBank": false,
    "right1": false,
    "right2": false,
    "right3": false,
    "right4": false,
    "right5": false,
    "rightBank": false,
    "rightGate": true,
    "shooterDiverter": true,
    "spinnerValue": undefined,
    "troughRelease": false,
    "upper2": false,
    "upper3": false,
    "upperEject": false,
    "upperMagnet": false,
  },
  "state": Object {
    "_name": "Machine19",
    "children": Array [
      Object {
        "_name": "DropBank20",
        "children": Array [],
        "name": "DropBank20",
        "ownOuts": undefined,
        "state": Object {},
      },
      Object {
        "_name": "DropBank21",
        "children": Array [],
        "name": "DropBank21",
        "ownOuts": undefined,
        "state": Object {},
      },
      Object {
        "_name": "DropBank22",
        "children": Array [],
        "name": "DropBank22",
        "ownOuts": undefined,
        "state": Object {},
      },
      Object {
        "_name": "DropBank23",
        "children": Array [],
        "name": "DropBank23",
        "ownOuts": undefined,
        "state": Object {},
      },
      Object {
        "_name": "DropBank24",
        "children": Array [],
        "name": "DropBank24",
        "ownOuts": undefined,
        "state": Object {},
      },
      Object {
        "_name": "DropBank25",
        "children": Array [],
        "name": "DropBank25",
        "ownOuts": undefined,
        "state": Object {},
      },
      Object {
        "_name": "Game29",
        "children": Array [
          Object {
            "_name": "Player30",
            "children": Array [
              Object {
                "_name": "ClearHoles31",
                "children": Array [],
                "name": "ClearHoles31",
                "ownOuts": Object {
                  "miniEject": false,
                  "outhole": false,
                  "upperEject": false,
                },
                "state": Object {},
              },
              Object {
                "_name": "Ball35",
                "children": Array [
                  Object {
                    "_name": "ResetAnyDropOnComplete36",
                    "children": Array [
                      Object {
                        "_name": "DropBankResetter37",
                        "children": Array [],
                        "name": "DropBankResetter37",
                        "ownOuts": Object {
                          "upper3": false,
                        },
                        "state": Object {},
                      },
                      Object {
                        "_name": "DropBankResetter38",
                        "children": Array [],
                        "name": "DropBankResetter38",
                        "ownOuts": Object {
                          "upper2": false,
                        },
                        "state": Object {},
                      },
                      Object {
                        "_name": "DropBankResetter39",
                        "children": Array [],
                        "name": "DropBankResetter39",
                        "ownOuts": Object {
                          "centerBank": false,
                        },
                        "state": Object {},
                      },
                      Object {
                        "_name": "DropBankResetter40",
                        "children": Array [],
                        "name": "DropBankResetter40",
                        "ownOuts": Object {
                          "miniBank": false,
                        },
                        "state": Object {},
                      },
                      Object {
                        "_name": "DropBankResetter41",
                        "children": Array [],
                        "name": "DropBankResetter41",
                        "ownOuts": Object {
                          "leftBank": false,
                        },
                        "state": Object {},
                      },
                      Object {
                        "_name": "DropBankResetter42",
                        "children": Array [],
                        "name": "DropBankResetter42",
                        "ownOuts": Object {
                          "rightBank": false,
                        },
                        "state": Object {},
                      },
                    ],
                    "name": "ResetAnyDropOnComplete36",
                    "ownOuts": undefined,
                    "state": Object {},
                  },
                ],
                "name": "Ball35",
                "ownOuts": Object {},
                "state": Object {
                  "skillshot": undefined,
                  "tilted": false,
                },
              },
              Object {
                "_name": "Spinner32",
                "children": Array [],
                "name": "Spinner32",
                "ownOuts": Object {
                  "iSpinner": undefined,
                  "leftGate": false,
                },
                "state": Object {
                  "comboMult": 1,
                  "rounds": 0,
                  "score": 800,
                },
              },
              Object {
                "_name": "LeftOrbit33",
                "children": Array [],
                "name": "LeftOrbit33",
                "ownOuts": Object {
                  "rightGate": true,
                },
                "state": Object {
                  "comboMult": 1,
                  "rounds": 1,
                  "score": 20000,
                },
              },
              Object {
                "_name": "Poker43",
                "children": Array [],
                "name": "Poker43",
                "ownOuts": Object {
                  "getSkillshot": [Function],
                  "iCenter1": undefined,
                  "iCenter2": undefined,
                  "iCenter3": undefined,
                  "iLeft1": undefined,
                  "iLeft2": undefined,
                  "iLeft3": undefined,
                  "iLeft4": undefined,
                  "iMini1": undefined,
                  "iMini2": undefined,
                  "iMini3": undefined,
                  "iRight1": undefined,
                  "iRight2": undefined,
                  "iRight3": undefined,
                  "iRight4": undefined,
                  "iRight5": undefined,
                  "iUpper21": undefined,
                  "iUpper22": undefined,
                  "iUpper31": undefined,
                  "iUpper32": undefined,
                  "iUpper33": undefined,
                  "lEjectShowCards": Array [
                    "White",
                  ],
                  "lRampShowCards": Array [
                    "White",
                  ],
                  "lShooterShowCards": Array [
                    "White",
                  ],
                  "lockPost": false,
                  "rampUp": false,
                  "shooterDiverter": true,
                  "upperEject": false,
                },
                "state": Object {
                  "closeShooter": false,
                  "dealerCardsUsed": Array [],
                  "dealerHand": Array [
                    Object {
                      "flipped": true,
                      "num": 3,
                      "suit": "h",
                    },
                    Object {
                      "flipped": true,
                      "num": 4,
                      "suit": "c",
                    },
                    Object {
                      "flipped": false,
                      "num": 9,
                      "suit": "h",
                    },
                    Object {
                      "flipped": false,
                      "num": 5,
                      "suit": "s",
                    },
                    Object {
                      "flipped": false,
                      "num": 9,
                      "suit": "d",
                    },
                    Object {
                      "flipped": false,
                      "num": 4,
                      "suit": "h",
                    },
                    Object {
                      "flipped": true,
                      "num": 2,
                      "suit": "c",
                    },
                  ],
                  "dealerHandDesc": "Pair",
                  "newMbs": Array [
                    Array [
                      "StraightMb",
                      Array [
                        Object {
                          "num": 7,
                          "suit": "s",
                        },
                        Object {
                          "num": 8,
                          "suit": "c",
                        },
                        Object {
                          "num": 9,
                          "suit": "c",
                        },
                        Object {
                          "num": 10,
                          "suit": "s",
                        },
                        Object {
                          "num": 11,
                          "suit": "s",
                        },
                      ],
                    ],
                  ],
                  "newModes": Array [],
                  "playerCardsUsed": Array [],
                  "playerHand": Array [
                    Object {
                      "num": 7,
                      "suit": "s",
                    },
                    Object {
                      "num": 11,
                      "suit": "s",
                    },
                    Object {
                      "num": 8,
                      "suit": "c",
                    },
                    Object {
                      "num": 9,
                      "suit": "c",
                    },
                    Object {
                      "num": 10,
                      "suit": "s",
                    },
                    Object {
                      "num": 13,
                      "suit": "s",
                    },
                    Object {
                      "num": 13,
                      "suit": "c",
                    },
                  ],
                  "playerHandDesc": "Straight",
                  "playerWins": undefined,
                  "pot": 1400,
                  "slots": Array [
                    Object {
                      "num": 1,
                      "suit": "d",
                    },
                    null,
                    null,
                    null,
                    Object {
                      "num": 1,
                      "suit": "s",
                    },
                    Object {
                      "num": 13,
                      "suit": "h",
                    },
                    Object {
                      "num": 3,
                      "suit": "d",
                    },
                    Object {
                      "num": 13,
                      "suit": "d",
                    },
                    null,
                    null,
                    Object {
                      "num": 11,
                      "suit": "h",
                    },
                    Object {
                      "num": 8,
                      "suit": "s",
                    },
                    Object {
                      "num": 10,
                      "suit": "h",
                    },
                    Object {
                      "num": 12,
                      "suit": "d",
                    },
                    Object {
                      "num": 8,
                      "suit": "d",
                    },
                    Object {
                      "num": 10,
                      "suit": "d",
                    },
                    Object {
                      "num": 12,
                      "suit": "s",
                    },
                    Object {
                      "num": 1,
                      "suit": "h",
                    },
                    Object {
                      "num": 9,
                      "suit": "s",
                    },
                    Object {
                      "num": 1,
                      "suit": "c",
                    },
                  ],
                  "step": 7,
                },
              },
              Object {
                "_name": "PlayerOverrides34",
                "children": Array [],
                "name": "PlayerOverrides34",
                "ownOuts": Object {
                  "kickerEnable": undefined,
                  "leftGate": undefined,
                  "miniFlipperEnable": undefined,
                  "rampUp": undefined,
                  "rightGate": undefined,
                  "shooterDiverter": undefined,
                },
                "state": Object {},
              },
            ],
            "name": "Player30",
            "ownOuts": Object {
              "lEjectStartMode": Array [],
              "lLaneUpper1": Array [
                "Orange",
              ],
              "lLaneUpper2": Array [
                "Orange",
              ],
              "lLaneUpper3": Array [
                "Orange",
              ],
              "lLaneUpper4": Array [
                "Orange",
              ],
              "lMiniReady": Array [
                "Red",
              ],
              "lPopperStatus": Array [
                "Green",
              ],
              "lPower1": Array [
                "Orange",
              ],
              "lPower2": Array [
                "Orange",
              ],
              "lPower3": Array [],
              "lPower4": Array [],
              "lRampMini": Array [
                "Orange",
              ],
              "lRampStartMb": Array [
                Array [
                  "Green",
                  "fl",
                ],
              ],
              "lShooterStartHand": Array [
                Array [
                  "Green",
                  "fl",
                ],
              ],
              "leftMagnet": false,
              "rampUp": false,
              "shooterDiverter": true,
            },
            "state": Object {
              "_score": 58000,
              "ball": "Ball35",
              "chips": 2,
              "closeShooter": false,
              "focus": "Poker43",
              "laneChips": Array [
                true,
                true,
                true,
                true,
              ],
              "mbsQualified": Array [],
              "miniReady": false,
              "modesQualified": Array [],
              "rampUp": true,
            },
          },
        ],
        "name": "Game29",
        "ownOuts": Object {
          "kickerEnable": true,
        },
        "state": Object {
          "ballNum": 1,
          "playerUp": 0,
          "players": Array [
            "Player30",
          ],
        },
      },
      Object {
        "_name": "MachineOverrides26",
        "children": Array [
          Object {
            "_name": "EosPulse27",
            "children": Array [],
            "name": "EosPulse27",
            "ownOuts": undefined,
            "state": Object {},
          },
        ],
        "name": "MachineOverrides26",
        "ownOuts": Object {
          "rampUp": undefined,
          "realRightBank": false,
          "shooterDiverter": undefined,
        },
        "state": Object {},
      },
    ],
    "name": "Machine19",
    "ownOuts": Object {
      "centerBank": false,
      "getSkillshot": undefined,
      "iCenter1": "",
      "iCenter2": "",
      "iCenter3": "",
      "iLeft1": "",
      "iLeft2": "",
      "iLeft3": "",
      "iLeft4": "",
      "iMini1": "",
      "iMini2": "",
      "iMini3": "",
      "iRight1": "",
      "iRight2": "",
      "iRight3": "",
      "iRight4": "",
      "iRight5": "",
      "iSS1": "",
      "iSS2": "",
      "iSS3": "",
      "iSS4": "",
      "iSS5": "",
      "iSS6": "",
      "iSpinner": "",
      "iUpper21": "",
      "iUpper22": "",
      "iUpper31": "",
      "iUpper32": "",
      "iUpper33": "",
      "ignoreSkillsot": Set {},
      "kickerEnable": false,
      "lEjectArrow": Array [],
      "lEjectShowCards": Array [],
      "lEjectStartMode": Array [],
      "lLaneUpper1": Array [],
      "lLaneUpper2": Array [],
      "lLaneUpper3": Array [],
      "lLaneUpper4": Array [],
      "lLeftArrow": Array [],
      "lLowerRamp": Array [],
      "lMagnet1": Array [
        "Green",
      ],
      "lMagnet2": Array [
        "Green",
      ],
      "lMagnet3": Array [],
      "lMagnet4": Array [],
      "lMainTargetArrow": Array [],
      "lMiniReady": Array [
        "Red",
      ],
      "lPopperStatus": Array [],
      "lPower1": Array [],
      "lPower2": Array [],
      "lPower3": Array [],
      "lPower4": Array [],
      "lRampArrow": Array [],
      "lRampMini": Array [],
      "lRampShowCards": Array [],
      "lRampStartMb": Array [],
      "lShooterLaneArrow": Array [],
      "lShooterShowCards": Array [],
      "lShooterStartHand": Array [],
      "lSideShotArrow": Array [],
      "lSideTargetArrow": Array [],
      "lSpinnerArrow": Array [],
      "lUpperLaneArrow": Array [],
      "lUpperTargetArrow": Array [],
      "leftBank": false,
      "leftGate": false,
      "leftMagnet": false,
      "lockPost": false,
      "magnetPost": false,
      "miniBank": false,
      "miniDiverter": false,
      "miniEject": false,
      "miniFlipperEnable": false,
      "outhole": false,
      "popper": false,
      "rampUp": false,
      "realRightBank": false,
      "right1": false,
      "right2": false,
      "right3": false,
      "right4": false,
      "right5": false,
      "rightBank": false,
      "rightGate": false,
      "shooterDiverter": false,
      "spinnerValue": undefined,
      "troughRelease": false,
      "upper2": false,
      "upper3": false,
      "upperEject": false,
      "upperMagnet": false,
    },
    "state": Object {
      "ballsInTrough": 3,
      "ballsLocked": 0,
      "lastSwitchHit": Switch {
        "_state": true,
        "column": 3,
        "lastChange": 30053,
        "lastClosed": 30053,
        "lastOpened": undefined,
        "minOffTime": 150,
        "minOnTime": 10,
        "name": "right 3",
        "row": 2,
      },
      "lockDown": false,
      "miniDown": false,
    },
  },
}
`;
